### PR TITLE
Better tetris demo

### PR DESCRIPTION
**Pull request by asokolsky**
Support:

* variable update rate
* more keys, e.g. to slow down/speed up fall rate, rotate in all directions, drop stone

### CODE REVIEW

**Comment by pvcraven** [comment]
Keep in mind with the demos, the idea is to make them as simple as possible. Highlight the new technique only. Let the reader use it as a springboard to make it fancy.

This does some things to improve docs and make simpler. It also expands the game a bit more. Are the expansions needed to show a new concept? If so, make sure the concept is clear. If not, keep it on the simple side.

**Comment by asokolsky** [comment]
@pvcraven understood and agreed.  I believe the PR does all that.  Plus allowed me to scratch my own itch, and for that opportunity I am grateful.  If you have specific points - I will be happy to address these.

Cheers!

File: arcade/examples/tetris.py [hunk 1]
```diff
@@ -11,6 +11,8 @@
 import arcade
 import random
 import PIL
+import pyglet
+import time
```
Comment (pvcraven) [inline arcade/examples/tetris.py:15]: For examples that demonstrate a certain concept, like how-to-do-Tetris I'd like to keep things narrow. Basic concept, as easy as possible to understand.

This should demo how to do a basic Tetris. If we want to center the window on the screen, adjust timings, etc. that should be a separate demo. (As these imports are related to.)

A full bells-and-whistles Tetris should be done as a different example. Probably in a different repository.

A step-by-step tutorial would be another, even more useful, option.

File: arcade/examples/tetris.py [hunk 2]
```diff
@@ -42,132 +41,208 @@
...
 
-    [[0, 0, 5],
+    [[0, 0, 5],         # Orange Ricky
      [5, 5, 5]],
 
-    [[6, 6, 6, 6]],
+    [[6, 6, 6, 6]],     # Hero
 
-    [[7, 7],
+    [[7, 7],            # Smashboy
```
Comment (pvcraven) [inline arcade/examples/tetris.py:62]: These updates are an improvement.

File: arcade/examples/tetris.py [hunk 3]
```diff
@@ -42,132 +41,208 @@
...
+        shape[y][x] for y in range(len(shape))]
+            for x in range(len(shape[0]) - 1, -1, -1)]
 
 
-def rotate_counterclockwise(shape):
+def rotate_clockwise(shape):
     """ Rotates a matrix clockwise """
-    return [[shape[y][x] for y in range(len(shape))]
-            for x in range(len(shape[0]) - 1, -1, -1)]
+    return [list(l[::-1]) for l in zip(*shape)]
```
Comment (pvcraven) [inline arcade/examples/tetris.py:76]: This would be an ok addition.

File: arcade/examples/tetris.py [hunk 4]
```diff
@@ -42,132 +41,208 @@
...
 
 
-def remove_row(board, row):
-    """ Remove a row from the board, add a blank row on top. """
-    del board[row]
-    return [[0 for _ in range(COLUMN_COUNT)]] + board
+#
+# List of images per color in colors
+#
+texture_list = [create_texture(color) for color in colors]
```
Comment (pvcraven) [inline arcade/examples/tetris.py:89]: This is nicer code. I do sit on the fence for tutorials using list comprehensions as I think they detract from readability. Although this is easier to read than the comprehensions for rotation.

File: arcade/examples/tetris.py [hunk 5]
```diff
@@ -42,132 +41,208 @@
...
+    def __init__(self, rows, columns, height):
+        """
+        Create a grid of 0's.
+        Add 1's to the bottom for easier collision detection.
+        """
+        self._rows = rows
+        self._columns = columns
+        self._height = height
+        self._mw = MARGIN + WIDTH
+        self._mh = MARGIN + HEIGHT
```
Comment (pvcraven) [inline arcade/examples/tetris.py:106]: For tutorials, should we spell these out?

File: arcade/examples/tetris.py [hunk 6]
```diff
@@ -42,132 +41,208 @@
...
 
 
-def join_matrixes(matrix_1, matrix_2, matrix_2_offset):
-    """ Copy matrix 2 onto matrix 1 based on the passed in x, y offset coordinate """
-    offset_x, offset_y = matrix_2_offset
-    for cy, row in enumerate(matrix_2):
-        for cx, val in enumerate(row):
-            matrix_1[cy + offset_y - 1][cx + offset_x] += val
-    return matrix_1
+class Board:
```
Comment (pvcraven) [inline arcade/examples/tetris.py:92]: This lengthens the code, significantly in what's supposed to be a simple how-to. Seem better for a longer example?

File: arcade/examples/tetris.py [hunk 7]
```diff
@@ -42,132 +41,208 @@
...
-
-        self.board = None
-        self.frame_count = 0
+        # Do the math to figure out our screen dimensions
+        width = (WIDTH + MARGIN) * cols + MARGIN
+        height = (HEIGHT + MARGIN) * ROW_COUNT + MARGIN
+
+        self._time_next_update = 0.0
+        self._desired_update_rate_per_sec = 3
+        self._desired_update_delay = 1/self._desired_update_rate_per_sec
```
Comment (pvcraven) [inline arcade/examples/tetris.py:194]: Variable timing should probably be its own example.

File: arcade/examples/tetris.py [hunk 8]
```diff
@@ -42,132 +41,208 @@
...
+        return
+
+    def center_the_window(self) -> None:
+        """Centers the window on the screen."""
+        SCREEN_NUM = 0
+        screen = pyglet.canvas.Display().get_screens()[SCREEN_NUM]
+        self.set_location(
+            screen.width // 2 - self.width // 2,
+            screen.height // 2 - self.height // 2)
+        return
```
Comment (pvcraven) [inline arcade/examples/tetris.py:225]: Center the window should be its own example.

File: arcade/examples/tetris.py [hunk 9]
```diff
@@ -176,105 +251,129 @@ def drop(self):
...
+            self.drop_stone()
+        elif key == arcade.key.SPACE:
+            self.drop_stone_hard()
+        elif key == arcade.key.P:
+            self.paused = not self.paused
+        elif key in (arcade.key.PLUS, arcade.key.NUM_ADD):
+            self.change_update_rate(1)
+        elif key in (arcade.key.MINUS, arcade.key.NUM_SUBTRACT):
+            self.change_update_rate(-1)
+        return
```
Comment (pvcraven) [inline arcade/examples/tetris.py:341]: This adds a lot of options. I'd suggest pushing to a larger stand-alone or a step-by-step tutorial.

**Review by pvcraven — CHANGES_REQUESTED** [review]
Really like to keep this example as simple as possible. Specific thoughts in line comments.

**Pull request closed without merging**